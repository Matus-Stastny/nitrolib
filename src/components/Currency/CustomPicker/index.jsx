// @flow strict
import * as React from "react";
import styled, { css } from "styled-components";

import mq from "../../../styles/mediaQuery";
import { themeDefault } from "../../../records/Theme";
import type { ThemeProps } from "../../../records/Theme";
import ClickOutside from "../../ClickOutside";
import type { Currency } from "../../../records/Currency";
import Current from "./Current";
import Menu from "./Menu";

const Container = styled.div`
  ${mq.gtDesktop(css`
    position: relative;
  `)};
`;

const OpenButton = styled.div`
  cursor: pointer;

  ${Container}:hover {
    color: ${({ theme }: ThemeProps) => theme.orbit.paletteProductNormal};
  }
`;

OpenButton.defaultProps = {
  theme: themeDefault,
};

type Props = {|
  current: Currency,
  available: Currency[],
  recommended: Currency[],
  onChange: (code: string) => void,
|};

type State = {|
  shown: boolean,
|};

export default class CustomPicker extends React.Component<Props, State> {
  state = {
    shown: false,
  };

  handleToggle = () => {
    this.setState(state => ({ shown: !state.shown }));
  };

  handleHide = () => {
    this.setState({ shown: false });
  };

  handleSetCurrency = (code: string) => {
    const { onChange } = this.props;
    this.setState({ shown: false });
    onChange(code);
  };

  render() {
    const { current, available, recommended } = this.props;
    const { shown } = this.state;

    return (
      <Container>
        <OpenButton onClick={this.handleToggle}>
          <Current currency={current} />
        </OpenButton>

        {shown && (
          <ClickOutside onClickOutside={this.handleHide}>
            <Menu
              current={current}
              available={available}
              recommended={recommended}
              onSetCurrency={this.handleSetCurrency}
            />
          </ClickOutside>
        )}
      </Container>
    );
  }
}
