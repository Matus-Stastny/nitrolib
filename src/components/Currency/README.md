# Currency

A currency picker.

**Selectors:**
* Open button: `[data-test="Currency"] button[data-test="Open"]`
* Item: `[data-test="Currency"] [data-test="eur"]`
