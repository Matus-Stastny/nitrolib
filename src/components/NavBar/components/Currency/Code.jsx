// @flow strict
import styled from "styled-components";

const Code = styled.span``;

export default Code;
