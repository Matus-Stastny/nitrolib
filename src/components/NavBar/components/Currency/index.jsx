// @flow strict
import * as React from "react";
import styled, { css } from "styled-components";
import { graphql, QueryRenderer } from "react-relay";

import Button from "../../primitives/Button";
import environmentReal from "services/environment";
import * as currencyContext from "services/currency/context";
import ClickOutside from "components/ClickOutside";
import mq from "styles/mediaQuery";
import Current from "./Current";
import Menu from "./Menu";

const Container = styled.div`
  ${mq.gtDesktop(css`
    position: relative;
  `)};
`;

type Props = {|
  // DI
  environment: typeof environmentReal,
|};

type State = {|
  shown: boolean,
|};

class Currency extends React.Component<Props, State> {
  static defaultProps = {
    environment: environmentReal,
  };

  state = {
    shown: false,
  };

  handleToggle = () => {
    this.setState(state => ({ shown: !state.shown }));
  };

  handleHide = () => {
    this.setState({ shown: false });
  };

  render() {
    const { environment } = this.props;
    const { shown } = this.state;

    return (
      <QueryRenderer
        query={graphql`
          query CurrencyQuery($ip: IP!) {
            currencies {
              ...MenuList_list
            }

            geoIP(ip: $ip) {
              ...MenuGeo_geo
            }
          }
        `}
        environment={environment}
        variables={{ ip: "1.2.3.4" /* TODO remove once 'geoIP' can exist without a variable */ }}
        render={res => {
          if (res.error) {
            // TODO
            console.error(res.error);
            return <span>Error :(</span>;
          }

          if (!res.props) {
            // TODO
            return null;
          }

          return (
            <currencyContext.Consumer>
              {({ currency, available, setCurrency }) => (
                <Container>
                  <Button onClick={this.handleToggle}>
                    <Current currency={currency} />
                  </Button>

                  {shown && (
                    <ClickOutside onClickOutside={this.handleHide}>
                      <Menu
                        current={currency}
                        available={available}
                        onSetCurrency={setCurrency}
                        onHide={this.handleHide}
                        /* $FlowIssue */
                        list={res.props.currencies}
                        /* $FlowIssue */
                        geo={res.props.geoIP}
                      />
                    </ClickOutside>
                  )}
                </Container>
              )}
            </currencyContext.Consumer>
          );
        }}
      />
    );
  }
}

export default Currency;
