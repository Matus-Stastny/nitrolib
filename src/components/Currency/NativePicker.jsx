// @flow strict
import * as React from "react";
import styled from "styled-components";
import * as R from "ramda";
import { Exchange } from "@kiwicom/orbit-components/lib/icons";

import NativeGroupedSelect from "../NativeGroupedSelect";
import { getCode, getSymbol } from "../../records/Currency";
import type { Currency } from "../../records/Currency";

const Container = styled.div`
  display: flex;
  flex-direction: row;
`;

const mapCurrencies = R.map(currency => ({
  value: currency.id,
  text: `${getCode(currency.id)} - ${getSymbol(currency.format)}`,
}));

type Props = {|
  current: Currency,
  available: Currency[],
  recommended: Currency[],
  onChange: (code: string) => void,
|};

const NativePicker = ({ current, available, recommended, onChange }: Props) => (
  <Container>
    <NativeGroupedSelect
      value={current.id}
      groups={[
        { key: "current", items: mapCurrencies([current]) },
        { key: "recommended", items: mapCurrencies(recommended) },
        { key: "all", items: mapCurrencies(available) },
      ]}
      icon={<Exchange className="CurrencySwitcher-icon" size="small" />}
      onChange={onChange}
    />
  </Container>
);

export default NativePicker;
