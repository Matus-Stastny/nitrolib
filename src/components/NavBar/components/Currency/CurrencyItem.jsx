// @flow strict
import * as React from "react";
import { graphql, createFragmentContainer } from "react-relay";
import styled, { css } from "styled-components";
import idx from "idx";

import { getCode, getSymbol } from "records/Currency";
import mq from "styles/mediaQuery";
import Code from "./Code";
import Sign from "./Sign";
import Name from "./Name";
import type { CurrencyItem_item } from "./__generated__/CurrencyItem_item.graphql";

const NameSeparator = styled.span`
  margin: 0 5px;

  ${mq.gtTablet(css`
    display: none;
  `)};
`;

type Props = {|
  item: CurrencyItem_item,
|};

const CurrencyItem = ({ item }: Props) => (
  <>
    <Code>{getCode(idx(item, _ => _.code) || "")}</Code>
    <Sign>{getSymbol(idx(item, _ => _.format) || "")}</Sign>
    <NameSeparator>-</NameSeparator>
    <Name>{idx(item, _ => _.name)}</Name>
  </>
);

export const CurrencyItemUnwrapped = CurrencyItem;

export default createFragmentContainer(
  CurrencyItem,
  graphql`
    fragment CurrencyItem_item on CurrencyDetail {
      code
      name
      format
    }
  `,
);
