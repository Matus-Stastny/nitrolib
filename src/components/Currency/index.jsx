// @flow strict
import * as React from "react";

import * as currencyContext from "../../services/currency/context";
import NativePicker from "./NativePicker";
import CustomPicker from "./CustomPicker";
import getAvailableList from "./services/getAvailableList";

type Props = {|
  native: boolean,
  loadingContent: React.Node,
|};

const Currency = ({ native, loadingContent }: Props) => (
  <currencyContext.Consumer>
    {({ currency, available, recommended, setCurrency }) => {
      if (!currency) {
        return loadingContent;
      }

      const availableList = getAvailableList(available);

      return native ? (
        <NativePicker
          current={currency}
          available={availableList}
          recommended={recommended}
          onChange={setCurrency}
        />
      ) : (
        <CustomPicker
          current={currency}
          available={availableList}
          recommended={recommended}
          onChange={setCurrency}
        />
      );
    }}
  </currencyContext.Consumer>
);

Currency.defaultProps = {
  native: false,
  loadingContent: null,
};

export default Currency;
