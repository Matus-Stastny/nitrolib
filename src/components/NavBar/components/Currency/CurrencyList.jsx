// @flow strict
import * as React from "react";
import styled, { css } from "styled-components";
import { graphql, createFragmentContainer } from "react-relay";

import Tooltip from "components/Tooltip";
import type { Currencies, Currency } from "records/Currency";
import { themeDefault } from "records/Theme";
import separateList from "services/utils/separateList";
import Flex from "primitives/Flex";
import mq from "styles/mediaQuery";
import CurrencyItem from "./CurrencyItem";
import Code from "./Code";
import Name from "./Name";
import Sign from "./Sign";
import type { CurrencyList_list } from "./__generated__/CurrencyList_list.graphql";

const COLUMNS = 4;

const Column = styled.div`
  width: 100%;

  ${mq.gtTablet(css`
    width: ${100 / COLUMNS}%;
  `)};
`;

const Item = styled.div`
  ${mq.gtTablet(css`
    padding-right: 25px;
  `)};
`;

const ItemText = styled.div`
  line-height: 30px;
  padding-left: 5px;
  border-radius: 3px;
  cursor: pointer;
  background: ${({ theme, active }) => (active ? theme.orbit.paletteProductNormal : "transparent")};

  &:hover {
    background: ${({ theme, active }) => theme.orbit[active ? "primary-600" : "grey-100"]};
  }

  ${Code} {
    font-weight: bold;
    color: ${({ theme, active }) => theme.orbit[active ? "white" : "grey-900"]};
  }

  ${Sign} {
    margin-left: 10px;
    color: ${({ theme, active }) => theme.orbit[active ? "white" : "grey-700"]};
  }

  ${Name} {
    margin-left: 10px;
    color: ${({ theme, active }) => theme.orbit[active ? "white" : "inherit"]};
  }

  ${mq.gtTablet(css`
    ${Name} {
      display: none;
    }
  `)};
`;

ItemText.defaultProps = {
  theme: themeDefault,
};

const Tip = styled.span`
  white-space: nowrap;
`;

type Props = {|
  // eslint-disable-next-line react/no-unused-prop-types
  active: Currency,
  available: Currencies,
  list: CurrencyList_list,
  // eslint-disable-next-line react/no-unused-prop-types
  onSetCurrency: (code: string) => void,
|};

const CurrencyList = ({ active, available, list, onSetCurrency }: Props) => (
  <Flex wrap="wrap">
    {separateList(
      COLUMNS,
      // FIXME filter directly in GraphQL
      (list.edges &&
        list.edges
          .map(edge => edge && edge.node)
          .filter(Boolean)
          .filter(node => available[String(node.code)])) ||
        [],
    ).map((items, i) => (
      // eslint-disable-next-line react/no-array-index-key
      <Column key={i}>
        {items.map(item => (
          <Item key={item.code}>
            <Tooltip position={i > 1 ? "left" : "right"} tip={<Tip>{item.name}</Tip>}>
              <ItemText
                active={item.code === active.id}
                onClick={() => onSetCurrency(item.code || "")}
              >
                {/* $FlowIssue */}
                <CurrencyItem item={item} />
              </ItemText>
            </Tooltip>
          </Item>
        ))}
      </Column>
    ))}
  </Flex>
);

export const CurrencyListUnwrapped = CurrencyList;

export default createFragmentContainer(
  CurrencyList,
  graphql`
    fragment CurrencyList_list on CurrencyDetailConnection {
      edges {
        node {
          code
          name
          ...CurrencyItem_item
        }
      }
    }
  `,
);
