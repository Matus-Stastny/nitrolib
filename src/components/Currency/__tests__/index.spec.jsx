// @flow strict
import * as React from "react";
import { shallow } from "enzyme";

import Currency from "../index";

describe("#Currency", () => {
  test("render", () => {
    const wrapper = shallow(<Currency />);

    expect(wrapper).toMatchSnapshot();
  });

  test("render - native", () => {
    const wrapper = shallow(<Currency native />);

    expect(wrapper).toMatchSnapshot();
  });
});
