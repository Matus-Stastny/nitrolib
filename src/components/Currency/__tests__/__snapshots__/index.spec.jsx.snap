// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Currency render - native 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`#Currency render 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
